ospf6d/ospf6_lsa.c

```diff
@@ -37,8 +38,256 @@ DEFINE_MTYPE_STATIC(OSPF6D, OSPF6_LSA,         "OSPF6 LSA");
+	char *end = ospf6_lsa_end(header);
+	int i;
+
+	if (ntohs(header->type) != OSPF6_LSTYPE_LINK &&
+	    ntohs(header->type) != OSPF6_LSTYPE_INTRA_PREFIX)
+		return NULL;
+
+	for (i = 0;
+	     i < pos && (char *)prefix + OSPF6_PREFIX_SIZE(prefix) < end; i++)
+		prefix = OSPF6_PREFIX_NEXT(prefix);
+
+	if (i == pos)
+		return prefix;
+
+	return NULL;
+}
+
+void *nth_tlv(struct ospf6_lsa_header *header, int pos)
+{
+	struct tlv_header *cur = lsdesc_start(header);
+	char *end = ospf6_lsa_end(header);
+	int i = 0;
+
+	for (; i < pos && (char *)cur + TLV_SIZE(cur) < end; i++)
+		cur = TLV_HDR_NEXT(cur);
+
+	if (i == pos)
+		return cur;
+
+	return NULL;
+}
+
+/* Router LSA */
+static int each_router_lsdesc(struct ospf6_lsa_header *lsa_header,
+			      struct tlv_handler *h)
+{
+	struct ospf6_router_lsdesc *desc = lsdesc_start(lsa_header);
+	char *lsa_end = ospf6_lsa_end(lsa_header);
+	int err = 0;
+
+	for (; (char *)(desc + 1) <= lsa_end && !err; desc++)
+		err = (*h->callback)(desc, h->callback_data);
+	return err;
+}
+
+/* Network LSA */
+static int each_network_lsdesc(struct ospf6_lsa_header *lsa_header,
+			       struct tlv_handler *h)
+{
+	struct ospf6_network_lsdesc *desc = lsdesc_start(lsa_header);
+	char *lsa_end = ospf6_lsa_end(lsa_header);
+	int err = 0;
+
+	for (; (char *)(desc + 1) <= lsa_end && !err; desc++)
+		err = (*h->callback)(desc, h->callback_data);
+	return err;
+}
+
+/* Link LSA */
+static int each_prefix_in_link_lsa(struct ospf6_lsa_header *lsa_header,
+				   struct tlv_handler *h)
+{
+	struct ospf6_prefix *prefix = lsdesc_start(lsa_header);
+	char *lsa_end = ospf6_lsa_end(lsa_header);
+	int err = 0;
+
+	for (; (char *)prefix + OSPF6_PREFIX_SIZE(prefix) <= lsa_end && !err;
+	     prefix = OSPF6_PREFIX_NEXT(prefix))
+		err = (*h->callback)(prefix, h->callback_data);
+	return err;
+}
+
+static inline int tlv_type_to_lsdesc_body_size(uint16_t tlv_type)
+{
+	return (tlv_type < OSPF6_TLV_ENUM_END)
+		       ? (sizeof(struct tlv_header) + tlv_size_map[tlv_type])
+		       : 0;
+}
+
+/* forward declaration */
+static int foreach_tlv(struct tlv_header *header, char *end,
+		       struct tlv_handler *tlvhandler);
+
+static int handle_one_tlv(struct tlv_header *header, struct tlv_handler *handler)
+{
+	int lsdesc_size = tlv_type_to_lsdesc_body_size(ntohs(header->type));
+	struct tlv_header *stlv_start;
+	char *tlv_end;
+	struct tlv_handler *sh;
+	int err;
+
+	assert(lsdesc_size);
+	err = (*handler->callback)(TLV_BODY(header), handler->callback_data);
+
+	if (err)
+		return err;
+	/* TODO: log something? */
+
+	/*
+	 * Are there any Sub-TLVs?
+	 * According to RFC 8362, "sub-TLVs may be nested to any level".
+	 * The sub-TLVs type IDs are mostly distinct in the E-LSA space,
+	 * but overlap with the TLV IDs, so they must be nested.
+	 *
+	 * FIXME: Calling the tlv iterator again creates an unbounded mutual
+	 * recursion, because the nesting is unbounded.
+	 */
+	if (TLV_SIZE(header) > lsdesc_size + TLV_HDR_SIZE) {
```

Comment:
This is trying to check if there are STLVs present based on whether the TLV is longer than the 'equivalent' LS descriptor record in a legacy LSA. It's wrong, because the LS desc size is variable - it depends on the prefix it carries.